### PR TITLE
Replace GitHub stubs with a fake

### PR DESCRIPTION
**Pull request by purinkle**
Before, all the calls to the GitHub API in the test environment were using WebMock stubs. This was adding lots of noise to the tests. Added a fake GitHub to replace the WebMock stubs.

* Removed redundant helper methods and fixtures.

### CODE REVIEW

**Comment by sardaukar** [comment]
I much prefer how this reads. Good one @purinkle

File: lib/github_api.rb [hunk 1]
```diff
@@ -5,6 +5,9 @@
 class GithubApi
   ORGANIZATION_TYPE = "Organization".freeze
 
+  cattr_accessor :client
+  self.client = Octokit::Client
```
Comment (gylaz) [inline lib/github_api.rb]: Why do this? What is the benefit? If we are overriding this somewhere, why not inject this into the constructor with a default?

File: spec/controllers/subscriptions_controller_spec.rb [hunk 2]
```diff
@@ -81,53 +81,57 @@
...
-      plan = tier.current.id
-      repo = membership.repo
-      token = "letmein"
-      stub_customer_find_request(user.stripe_customer_id)
-      stub_sign_in(user)
-      stub_hook_creation_request(repo.name, hook_url, token)
-      stub_subscription_create_request(plan: plan, repo_ids: repo.id)
-      stub_subscription_update_request(plan: new_plan, repo_ids: repo.id)
+      repo = instance_double(
+        Repo,
```
Comment (gylaz) [inline spec/controllers/subscriptions_controller_spec.rb]: If we use the string version of this (i.e. `instance_double("Repo"...)`) then the constant does not need to be loaded. That's what we prefer.

File: spec/controllers/subscriptions_controller_spec.rb [hunk 3]
```diff
@@ -81,53 +81,57 @@
...
+      user = instance_double(
+        User,
+        email: "TEST_USER_EMAIL",
+        repos: repos,
+        token: "TEST_USER_TOKEN",
+      )
+      users = class_double(User, first: user)
+      allow(RepoActivator).to receive(:new).and_return(repo_activator)
+      allow(RepoSubscriber).to receive(:subscribe).and_return(repo_subscriber)
+      allow(User).to receive(:where).and_return(users)
```
Comment (gylaz) [inline spec/controllers/subscriptions_controller_spec.rb]: This test does a lot of stubbing, but does it provide any value? Does the feature spec (if there is one) cover this code and the ability to update a subscription? Maybe we should just remove it...

File: spec/lib/github_api_spec.rb [hunk 4]
```diff
@@ -4,7 +4,12 @@
 require "json"
 
 describe GithubApi do
-  before { stub_const("Hound::GITHUB_TOKEN", "token") }
+  before do
+    stub_const("Hound::GITHUB_TOKEN", "token")
+    GithubApi.client = Octokit::Client
+  end
+
+  after { GithubApi.client = FakeGithub }
```
Comment (gylaz) [inline spec/lib/github_api_spec.rb]: Why do we set the client to `FakeGithub` after the test?

File: spec/requests/builds_spec.rb [hunk 5]
```diff
@@ -12,95 +12,36 @@
...
-        file_path: filename,
-        commit_id: commit_id_from_fixture,
-      )
 
       post builds_path, payload: payload
 
-      expect(new_comment_request).to have_been_requested
+      comments = FakeGithub.comments
+      comment = comments.first
+      expect(comments.length).to eq 1
```
Comment (gylaz) [inline spec/requests/builds_spec.rb]: `length` -> `size`.

File: spec/requests/builds_spec.rb [hunk 6]
```diff
@@ -12,95 +12,36 @@
...
+      expect(comment.repo).to eq repo_name
+      expect(comment.pull_id).to eq pr_number
+      expect(comment.body).to eq "Trailing whitespace detected."
+      expect(comment.commit_id).to eq pr_sha
+      expect(comment.path).to eq "spec/models/style_guide_spec.rb"
+      expect(comment.position).to eq 9
```
Comment (gylaz) [inline spec/requests/builds_spec.rb]: Can we use the `have_attributes` matcher for these?

File: spec/support/fake_github.rb [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
++
++after_fork do |server, worker|
++  Signal.trap 'TERM' do
++    puts 'Wait for master to send QUIT'
++  end
++
++  defined?(ActiveRecord::Base) and
++    ActiveRecord::Base.establish_connection
++end
+      PATCH
```
Comment (gylaz) [inline spec/support/fake_github.rb]: How about extracting this heredoc to a fixture?

File: .ruby-style.yml [hunk 8]
```diff
@@ -573,6 +573,8 @@ Rails/HasAndBelongsToMany:
   Enabled: true
   Include:
   - app/models/**/*.rb
+Rails/HttpPositionalArguments:
```
Comment (teoljungberg) [inline .ruby-style.yml]: Thoughts on extracting this to a separate PR?

File: lib/github_api.rb (hunk 1 above)
Comment (teoljungberg) [inline lib/github_api.rb]: I'd 👍 using dependency injection here instead. (which is also more thread-safe)

File: lib/github_api.rb [hunk 9]
```diff
@@ -13,7 +16,7 @@ def initialize(token)
   end
 
   def client
-    @client ||= Octokit::Client.new(access_token: token, auto_paginate: true)
+    @client ||= self.class.client.new(access_token: token, auto_paginate: true)
```
Comment (teoljungberg) [inline lib/github_api.rb]: Thoughts on using memoization with an underscore for consistency with the rest of the app? Or more specifically the new parts of the app?

File: spec/controllers/sessions_controller_spec.rb [hunk 10]
```diff
@@ -4,12 +4,16 @@
...
+        github_api = instance_double(
+          GithubApi,
+          scopes: "public_repo,user:email",
+        )
         request.env["omniauth.auth"] = stub_oauth(
           username: "jimtom",
           email: "jimtom@example.com",
           token: "letmein",
         )
+        allow(GithubApi).to receive(:new).and_return(github_api)
```
Comment (teoljungberg) [inline spec/controllers/sessions_controller_spec.rb]: Shouldn't this use the DI pattern, or use the `cattr_attribute`?

File: spec/controllers/subscriptions_controller_spec.rb (hunk 3 above)
Comment (teoljungberg) [inline spec/controllers/subscriptions_controller_spec.rb]: I agree, I normally find controller specs not to add alot of use. I lean on feature specs for when a user traverses an application, or request specs when it is an API.

File: spec/lib/github_api_spec.rb (hunk 4 above)
Comment (teoljungberg) [inline spec/lib/github_api_spec.rb]: It looks like we need to do it to get the state back on track after the tests. The `rails_helper` is setting `GithubApi.client = FakeGithub`, and that state needs to be reset.

Comment (teoljungberg) [inline spec/lib/github_api_spec.rb]: Could this be an `around` filter instead?

```ruby
around do |example|
  stub_const("Hound::GITHUB_TOKEN", "token")
  GithubApi.client = Octokit::Client
  
  example.run

  GithubApi.client = FakeGithub
end
```

File: spec/requests/builds_spec.rb (hunk 6 above)
Comment (teoljungberg) [inline spec/requests/builds_spec.rb]: I like that too!

File: spec/support/fake_github.rb [hunk 11]
```diff
@@ -0,0 +1,99 @@
+class FakeGithub
+  Comment = Struct.new(:repo, :pull_id, :body, :commit_id, :path, :position)
+  Content = Struct.new(:content)
+  File = Struct.new(:filename, :patch, :status)
+  Hook = Struct.new(:id)
+
+  cattr_accessor :comments
```
Comment (teoljungberg) [inline spec/support/fake_github.rb]: I generally try to shy away from `cattr_accessor`, as it's generally not thread-safe.

File: spec/support/fake_github.rb [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    self.class.comments << Comment.new(*args)
+  end
+
+  def pull_request_comments(repo, number)
+    [
+      Comment.new(
+        repo,
+        number,
+        "Line is too long.",
+        "81ee93414c230f71603265bd791bff5fbb8dba3e",
```
Comment (teoljungberg) [inline spec/support/fake_github.rb]: Can this SHA be a bit more generic?

File: spec/support/fake_github.rb [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+        "spec/models/style_guide_spec.rb",
```
Comment (teoljungberg) [inline spec/support/fake_github.rb]: Same as this filename.

File: spec/support/fake_github.rb [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+        5,
+      ),
+    ]
+  end
+
+  def pull_request_files(*_args)
+    [
+      File.new("spec/models/style_guide_spec.rb", <<-PATCH, "added")
+@@ -0,0 +1,22 @@
++worker_processes Integer(ENV[\"WEB_CONCURRENCY\"] || 3)
```
Comment (teoljungberg) [inline spec/support/fake_github.rb]: Same as this filecontents.

File: spec/support/fake_github.rb (hunk 7 above)
Comment (teoljungberg) [inline spec/support/fake_github.rb]: I like that too.

File: spec/support/fake_github.rb [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+  end
+
+  def remove_hook(*_args)
+    true
+  end
+
+  def repos
+    [
+      {
+        full_name: "jimtom/My-Private-Repo",
```
Comment (teoljungberg) [inline spec/support/fake_github.rb]: Can we have a more anonymous name?

File: spec/support/fake_github.rb [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+        id: 1296269,
```
Comment (teoljungberg) [inline spec/support/fake_github.rb]: Does this id need to be this specifically, or can it be a generic one such as `1`?

File: spec/controllers/sessions_controller_spec.rb (hunk 10 above)
Comment (purinkle) [inline spec/controllers/sessions_controller_spec.rb]: For this level of testing, we do not want the test to drop into the lower classes. The test should only care that `GithubApi` is being called and not what `GithubApi` is doing. I would argue that we should not be accessing `GithubApi` direct from the controller. This is a story for another day.

Comment (teoljungberg) [inline spec/controllers/sessions_controller_spec.rb]: Sounds correct to me!

**Comment by teoljungberg** [comment]
I like this alot! Great work @purinkle 

I want to hear what @gylaz thinks about this Sinatra-fake approach. I for one like it alot better

**Pull request merged**